### ISSUE TITLE
Future of Elm tooling in VSCode

### ISSUE THREAD

**Issue by Krzysztof-Cieslak** [Krzysztof-Cieslak/vscode-elm#334]
Hey everyone,

1. I think it's become pretty obvious for everyone that I don't really have time to maintain this project any more - I'm involved with way too many OSS projects in the ecosystem (F#) that's more important to me, and I'm not using Elm neither professionally nor as a hobby any more.
2. Any previous attempts to find additional maintainers for this repository failed - not blaming anyone, I know how hard it is to find enough time to do OSS
3. I'm huge believer in Language Server Protocol and I'm supper happy to see that Elm is getting language server using this protocol -  https://github.com/elm-tooling/elm-language-server

So... I think I'd want to somehow deprecate this project in favor of LSP based VSCode plugin (https://github.com/elm-tooling/elm-language-client-vscode). However `vscode-elm` is way more popular according to the VSCode download/installation stats, so we may think about some nice way of migrating. 
For example we may move `vscode-elm` plugin to the `elm-tooling` organization, and move code from `elm-language-client-vscode` to `vscode-elm` - thanks to that we'd be able to deliver the new LSP based tooling to all people using `vscode-elm` at the moment.

I'd be happy to hear opinions and suggestions

CC: @Razzeee

**Comment by Razzeee** [comment]
Thank you for taking the initiative on this.

At the moment I'm very focused on writing language server only features. Which means the vscode client is basic in regards to vscode only features.

My biggest question would be, if we remove features that people are using. For e.g. the language server is 0.19.0 only. But still, you could just install an old version of the extension.

In general, I think this is the right move. Still it would be nice to find somebody that really cares about the vscode implementation, so that I can focus on lsp. Or the other way around.

**Comment by Krzysztof-Cieslak** [comment]
> My biggest question would be, if we remove features that people are using. For e.g. the language server is 0.19.0 only. But still, you could just install an old version of the extension.

Ofc, breaking people is always sad, but TBF, current state of `vscode-elm` is not optimistic - I don't see myself or anyone else working on the editor features implementations and improving them. I'd rather take step back, even "sacrifice" some of the features temporarily but go with something that has future ahead - and from my point of view it's LSP.
Also, just from quick playing with it, it seems that editor features (tooltips, autocomplete, etc) are working well in LSP based plugin, so I'm not even sure if that's something we need to worry about.

Support for `0.18` is probably one huge thing that'll be missing... but again, I'd rather look forward. If someone is using `0.18` & VSCode then they haven't been getting any tooling updates for a long time anyway - they can keep using current version of `vscode-elm`.

> Which means the vscode client is basic in regards to vscode only features.

The VSCode-only features in `vscode-elm` are just bunch of commands that we can easily add to LSP plugin. Those are mostly commands allowing to run from editor things like:
* Elm REPL
* Elm Reactor 
* Elm Package
* Elm Make

> In general, I think this is the right move. Still it would be nice to find somebody that really cares about the vscode implementation, so that I can focus on lsp. Or the other way around.

I think I may help in initial stage, but in longer term I can't see myself being main maintainer of the Elm Code plugin. But one step at the time, I'd rather figure out for now what's the plan we want to follow going forward with that. 

My suggestion would be:
1. Move `vscode-elm` to `elm-tooling` organization
2. Remove all editor features implementations from `vscode-elm` keeping only custom commands I've mentioned earlier
3. Move LSP client implementation from `elm-language-client-vscode` to `vscode-elm`
4. Release new version of `vscode-elm` based on LSP

**Comment by andys8** [comment]
I really like the idea, too.

The first release with LSP should be well tested for some time. Currently there are still some issues here and there because it's still early.

In the long term it makes sense for all parties to focus on LSP. It's also the right step, because even if the current plugin isn't maintained anymore or is broken, people will install it from withing VSCode because of its rating. Since most elm users and new users use vs code this has an impact on how people will see Elm itself.

I myself would like to help, too. But a) I'm not a VSCode user, and b) @razzeee is doing all the work, has the knowledge about and deserves all the credit for the language server implementation :) I'm happy to help with the migration, alpha testing and so on.


As part of the migration we could also republish the current vs code plugin under a different name. People preferring or relying on it, could still use and maintain it. This could be probably easier than pinning an older version.

Is it possible to extract "editor features" to a separate plugin? What features do exist? (3.)

**Comment by Razzeee** [comment]
Btw, thanks for your work on Ionide, only using it for our F# Meetup, but still awesome extension 
 
> > Which means the vscode client is basic in regards to vscode only features.
> 
> The VSCode-only features in `vscode-elm` are just bunch of commands that we can easily add to LSP plugin. Those are mostly commands allowing to run from editor things like:
> 
>     * Elm REPL
> 
>     * Elm Reactor
> 
>     * Elm Package
> 
>     * Elm Make
> 

Yes, these were the ones I was thinking about, we might not even need all. We might even be able to move some to the lsp implementation. The only thing I'm not sure about are open ended processes like REPL and I think each client still needs to configure commands?
I only toyed with that api briefly.

> > In general, I think this is the right move. Still it would be nice to find somebody that really cares about the vscode implementation, so that I can focus on lsp. Or the other way around.
> 
> I think I may help in initial stage, but in longer term I can't see myself being main maintainer of the Elm Code plugin. But one step at the time, I'd rather figure out for now what's the plan we want to follow going forward with that.

Understandable, I did not mean to suggest you. Still exciting, that you would actually help out :)

> My suggestion would be:
> 
>     1. Move `vscode-elm` to `elm-tooling` organization
> 
>     2. Remove all editor features implementations from `vscode-elm` keeping only custom commands I've mentioned earlier
> 
>     3. Move LSP client implementation from `elm-language-client-vscode` to `vscode-elm`
> 
>     4. Release new version of `vscode-elm` based on LSP

In my mind it makes more sense to migrate the features from `vscode-elm` to `elm-language-client-vscode` and then rename the package (and maybe the repo to `vscode-elm`)

So that we're:
- Not getting the issues / PRs from the old repo
- No inheriting the complete git history and were more explicit which code gets migrated
- Keeping your repo (maybe archived) for reference / people that are curious what happened.

**Comment by Krzysztof-Cieslak** [comment]
> In my mind it makes more sense to migrate the features from vscode-elm to elm-language-client-vscode and then rename the package (and maybe the repo to vscode-elm)

Sounds good to me

**Comment by Razzeee** [comment]
I guess we should coordinate the next steps in slack or what do you think?
Maybe in the language server channel?

**Comment by hakonrossebo** [comment]
I agree with the proposed approach. I could help with implementing the VSCode-only features as I've been involved with that part. By using this approach, do I understand correctly that in the end, for end-users, the vscode-elm plugin everyone uses now will be updated to use lsp?

I also think moving the old vscode-elm plugin to a vscode-elm-old repo/plugin would make sense.

**Comment by Razzeee** [comment]
Yes, that's the idea. Minimal user interaction needed (unless you need 0.18.0).

**Comment by Krzysztof-Cieslak** [comment]
> 1. Move `vscode-elm` to `elm-tooling` organization

Step 1 done, we're now in `elm-tooling`

**Comment by Razzeee** [comment]
So how do we proceed?
